src/circuit/client.ts

```diff
@@ -1,57 +1,100 @@
 import { logger } from '@libp2p/logger'
-import { fromString as uint8ArrayFromString } from 'uint8arrays/from-string'
-import { toString as uint8ArrayToString } from 'uint8arrays/to-string'
-import { RELAY_CODEC } from './multicodec.js'
-import { canHop } from './circuit/hop.js'
-import { namespaceToCid } from './utils.js'
+import { RELAY_V2_HOP_CODEC } from './multicodec.js'
+import { getExpiration, namespaceToCid } from './utils.js'
 import {
   CIRCUIT_PROTO_CODE,
-  HOP_METADATA_KEY,
-  HOP_METADATA_VALUE,
+  DEFAULT_MAX_RESERVATIONS,
   RELAY_RENDEZVOUS_NS
 } from './constants.js'
 import type { PeerId } from '@libp2p/interface-peer-id'
-import type { AddressSorter, PeerProtocolsChangeData } from '@libp2p/interface-peer-store'
+import type { AddressSorter } from '@libp2p/interface-peer-store'
 import type { Connection } from '@libp2p/interface-connection'
 import sort from 'it-sort'
 import all from 'it-all'
 import { pipe } from 'it-pipe'
 import { publicAddressesFirst } from '@libp2p/utils/address-sort'
-import type { RelayComponents } from './index.js'
+import { reserve } from './v2/index.js'
+import { EventEmitter, CustomEvent } from '@libp2p/interfaces/events'
+import type { Startable } from '@libp2p/interfaces/startable'
+import type { Components } from '../components.js'
+import type { CircuitServiceConfig } from './index.js'
 
-const log = logger('libp2p:auto-relay')
+const log = logger('libp2p:circuit:client')
 
-const noop = () => {}
+const noop = () => { }
 
-export interface AutoRelayInit {
+/**
+ * CircuitServiceInit initializes the circuit service using values
+ * from the provided config and an @type{AddressSorter}.
+ */
+export interface CircuitServiceInit extends CircuitServiceConfig {
+  /**
+   * Allows prioritizing addresses from the peerstore for dialing. The
+   * default behavior is to prioritise public addresses.
+   */
   addressSorter?: AddressSorter
-  maxListeners?: number
+  /**
+   * A callback to invoke when an error occurs in the circuit service.
+   */
   onError?: (error: Error, msg?: string) => void
 }
 
-export class AutoRelay {
-  private readonly components: RelayComponents
+export interface CircuitServiceEvents {
```

Comment:
@mpetrunic I feel `AutoRelay` is a better name for this class than `CircuitService`. `CircuitService` feels like it should provide the relay service, but in this case it automatically connects to relays. Would `AutoReserve` be a better name?

Reply:
RelayReservationManager?